txstratum/protocol.py

```diff
@@ -0,0 +1,351 @@
+"""
+Copyright (c) Hathor Labs and its affiliates.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+"""
+
+import asyncio
+import time
+import uuid
+from functools import reduce
+from math import log2
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, NamedTuple, Optional, cast
+
+from structlog import get_logger  # type: ignore
+from txstratum.commons import Block, Transaction, sum_weights
+from txstratum.commons.utils import InvalidAddress, decode_address
+from txstratum.jobs import MinerBlockJob
+from txstratum.utils import JSONRPCError, JSONRPCId, JSONRPCProtocol, MaxSizeOrderedDict, Periodic
+
+if TYPE_CHECKING:
+    from txstratum.commons import BaseTransaction
+    from txstratum.jobs import MinerJob
+    from txstratum.manager import TxMiningManager  # noqa: F401
+
+logger = get_logger()
+
+
+class SubmittedWork(NamedTuple):
+    """Submitted work by the miner."""
+
+    timestamp: float
+    weight: float
+
+
+class StratumProtocol(JSONRPCProtocol):
+    """Implementation of Stratum Protocol.
+
+    Specification: https://en.bitcoin.it/wiki/Stratum_mining_protocol
+    """
+
+    ESTIMATOR_LOOP_INTERVAL = 30  # in seconds, "frequency" that the function that updates the estimator will be called
+    ESTIMATOR_WINDOW_INTERVAL = 60 * 15  # in seconds, size of the window to use for estimating miner's hashrate
+    TARGET_JOB_TIME = 15  # in seconds, adjust difficulty so jobs take this long
+
+    MIN_WEIGHT = 20  # minimum "difficulty" to assign to jobs
+    MAX_WEIGHT = 60  # maximum "difficulty" to assign to jobs
+    INITIAL_WEIGHT = 30  # initial "difficulty" to assign to jobs, can raise or drop based on solvetimes
+    MAX_JOBS = 1000  # maximum number of jobs to keep in memory
+
+    INVALID_ADDRESS = JSONRPCError(22, 'Address to send mined funds is invalid')
+    INVALID_SOLUTION = JSONRPCError(30, 'Invalid solution')
+    STALE_JOB = JSONRPCError(31, 'Stale job submitted')
+    JOB_NOT_FOUND = JSONRPCError(32, 'Job not found')
+
+    def __init__(self, manager: 'TxMiningManager'):
+        """Init StratumProcol to receive a new connection."""
+        self.log = logger.new()
+        self.manager: 'TxMiningManager' = manager
+        self.miner_id: str = str(uuid.uuid4())
+        self.miner_address: Optional[bytes] = None
+        self.miner_address_str: Optional[str] = None
+
+        self.jobs: MaxSizeOrderedDict[bytes, 'MinerJob'] = MaxSizeOrderedDict(max=self.MAX_JOBS)
+        self.current_job: Optional['MinerJob'] = None
+        self.current_weight: float = self.INITIAL_WEIGHT
+        self.hashrate_ghs: float = 0.0
+
+        self._submitted_work: List[SubmittedWork] = []
+
+        self._authorized: bool = False
+        self._subscribed: bool = False
+        self.started_at: float = 0.0
+        self.estimator_task: Optional[Periodic] = None
+
+        # For testing only.
+        self._update_job_timestamp: bool = True
```

Comment:
Why is this just for testing?

Reply:
I disable updating job timestamps to make testing easier. The issue was comparing two jobs there are supposed to be the same but they have different timestamps. But maybe I can use the hash of the funds when comparing. What do you think?

Reply:
Sorry, my explanation wasn't clear. I meant that I disabled updating job timestamp for some of the unittests.

Reply:
Ah, got it.